### ISSUE TITLE
Unit test with an io::Write that returns Ok(0)

### ISSUE THREAD

**Issue by dtolnay** [serde-rs/json#179]
This would have caught #178. Let's run all unit tests also against a writer that does not always succeed.

**Comment by oli-obk** [comment]
Can't we just mod the normal writer to always do that? It's not like one can write code accidentally that only works in the presence of paritial writes

**Comment by dtolnay** [comment]
Yeah good call. I was thinking to_string and to_writer were different codepaths but that's only for deserialization.
